File: a2ml-templates/state-scm-to-v2.py
#!/usr/bin/env python3
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# state_scm_to_v2.py — convert a v1 STATE.scm (s-expr machine state) to the
# canonical STATE.a2ml v2 "thin session journal" directive format.
#
# Spec: standards/a2ml-templates/STATE.a2ml.v2.spec.adoc
# v2 deliberately discards everything derivable; keeps only:
#   phase, next_action, last_action, updated, and @blockers entries.
#
# Usage:
#   state_scm_to_v2.py <STATE.scm>            # print v2 to stdout (dry-run)
#   state_scm_to_v2.py <STATE.scm> --write    # write .machine_readable/STATE.a2ml
#
# Exit codes: 0 ok, 2 parse/usage error, 3 not a (state ...) form.

import sys, os, re, datetime

def tokenize(s):
    # Strip ;; line comments, then tokenize parens and "strings" and atoms.
    s = re.sub(r';;[^\n]*', '', s)
    toks, i, n = [], 0, len(s)
    while i < n:
        c = s[i]
        if c in '()':
            toks.append(c); i += 1
        elif c == '"':
            j = i + 1; buf = []
            while j < n and s[j] != '"':
                if s[j] == '\\' and j + 1 < n:
                    buf.append(s[j+1]); j += 2
                else:
                    buf.append(s[j]); j += 1
            toks.append(('str', ''.join(buf))); i = j + 1
        elif c.isspace():
            i += 1
        else:
            j = i
            while j < n and not s[j].isspace() and s[j] not in '()"':
                j += 1
            toks.append(('atom', s[i:j])); i = j
    return toks

def parse(toks):
    # Returns nested lists; strings -> ('s', value), atoms -> ('a', value).
    pos = 0
    def rd():
        nonlocal pos
        t = toks[pos]; pos += 1
        if t == '(':
            lst = []
            while toks[pos] != ')':
                lst.append(rd())
            pos += 1
            return lst
        if t == ')':
            raise ValueError('unexpected )')
        return t  # ('str',..) or ('atom',..)
    forms = []
    while pos < len(toks):
        forms.append(rd())
    return forms

def head(node):
    return node[0][1] if (isinstance(node, list) and node and isinstance(node[0], tuple)) else None

def find(node, name):
    """First child list whose head atom == name."""
    if not isinstance(node, list):
        return None
    for ch in node:
        if isinstance(ch, list) and ch and isinstance(ch[0], tuple) and ch[0][1] == name:
            return ch
    return None

def first_string(node):
    """First string literal anywhere under node (depth-first)."""
    if isinstance(node, tuple):
        return node[1] if node[0] == 'str' else None
    if isinstance(node, list):
        for ch in node:
            r = first_string(ch)
            if r:
                return r
    return None

def all_strings(node):
    out = []
    if isinstance(node, tuple):
        if node[0] == 'str':
            out.append(node[1])
    elif isinstance(node, list):
        for ch in node:
            out.extend(all_strings(ch))
    return out

def q(s):
    return '"' + str(s).replace('\\', '\\\\').replace('"', '\\"') + '"'

def slug(s):
    return re.sub(r'[^a-z0-9]+', '-', s.lower()).strip('-')[:40] or 'blocker'

def convert(path):
    src = open(path, encoding='utf-8', errors='replace').read()
    forms = parse(tokenize(src))
    state = next((f for f in forms if head(f) == 'state'), None)
    if state is None:
        sys.stderr.write(f"ERR: no (state ...) form in {path}\n")
        sys.exit(3)

    # phase  <- (current-position (phase "..")) | (current-phase "..")
    phase = None
    cp = find(state, 'current-position')
    if cp:
        ph = find(cp, 'phase')
        if ph:
            phase = first_string(ph)
    if not phase:
        cph = find(state, 'current-phase')
        if cph:
            phase = first_string(cph)
    phase = phase or 'unknown'

    # next_action <- critical-next-actions, first string (immediate first)
    next_action = 'TODO — review and set'
    cna = find(state, 'critical-next-actions')
    if cna:
        imm = find(cna, 'immediate')
        next_action = first_string(imm) or first_string(cna) or next_action

    # last_action <- last session-history entry's first accomplishment string
    last_action = 'migrated from v1 STATE.scm'
    sh = find(state, 'session-history')
    if sh:
        sessions = [c for c in sh if isinstance(c, list) and head(c) == 'session']
        if sessions:
            acc = find(sessions[-1], 'accomplishments')
            last_action = first_string(acc) or first_string(sessions[-1]) or last_action

    # updated <- (metadata (updated|last-updated "..")) else today
    updated = None
    md = find(state, 'metadata')
    if md:
        for key in ('updated', 'last-updated'):
            u = find(md, key)
            if u:
                updated = first_string(u)
                break
    updated = updated or datetime.date.today().isoformat()

    # blockers <- (blockers-and-issues (critical ..)(high ..)(medium ..)(low ..))
    blockers = []
    bi = find(state, 'blockers-and-issues')
    if bi:
        for sev in ('critical', 'high', 'medium', 'low'):
            grp = find(bi, sev)
            if grp:
                for desc in all_strings(grp):
                    blockers.append((sev, desc))

    lines = []
    lines.append('# SPDX-License-Identifier: PMPL-1.0-or-later')
    lines.append(f'# Migrated from {os.path.basename(path)} by state_scm_to_v2.py on '
                 f'{datetime.date.today().isoformat()}')
    lines.append('')
    lines.append('@state(version="2.0"):')
    lines.append(f'phase: {q(phase)}')
    lines.append(f'next_action: {q(next_action)}')
    lines.append(f'last_action: {q(last_action)}')
    lines.append(f'updated: {updated}')
    if blockers:
        lines.append('')
        lines.append('@blockers:')
        for sev, desc in blockers:
            lines.append(f'- id: {slug(desc)}')
            lines.append(f'  description: {q(desc)}')
            lines.append(f'  waiting_on: {q(sev + "-priority — internal")}')
            lines.append(f'  since: {updated}')
        lines.append('@end')
    lines.append('')
    lines.append('@end')
    return '\n'.join(lines) + '\n'

def main():
    args = [a for a in sys.argv[1:] if a != '--write']
    write = '--write' in sys.argv
    if len(args) != 1:
        sys.stderr.write("usage: state_scm_to_v2.py <STATE.scm> [--write]\n")
        sys.exit(2)
    scm = args[0]
    out = convert(scm)
    if write:
        dest = os.path.join(os.path.dirname(scm), 'STATE.a2ml')
        open(dest, 'w', encoding='utf-8').write(out)
        sys.stderr.write(f"wrote {dest}\n")
    else:
        sys.stdout.write(out)

if __name__ == '__main__':
    main()

File: a2ml-templates/state-scm-to-v2.jl
#!/usr/bin/env julia
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# state-scm-to-v2.jl — convert a v1 STATE.scm (s-expr machine state) to the
# canonical STATE.a2ml v2 "thin session journal" directive format.
#
# Spec: standards/a2ml-templates/STATE.a2ml.v2.spec.adoc
# v2 deliberately discards everything derivable; keeps only:
#   phase, next_action, last_action, updated, and @blockers entries.
#
# Usage:
#   state-scm-to-v2.jl <STATE.scm>            # print v2 to stdout (dry-run)
#   state-scm-to-v2.jl <STATE.scm> --write    # write .machine_readable/STATE.a2ml
#
# Exit codes: 0 ok, 2 parse/usage error, 3 not a (state ...) form.
#
# Ported from the original Python implementation (Python is banned estate-wide);
# behaviour and output are identical. A node is either a Vector (a list) or a
# leaf tuple (:str, value) / (:atom, value).

using Dates

# --- s-expression reader ----------------------------------------------------

function tokenize(s::AbstractString)
    # Strip ;; line comments, then tokenize parens and "strings" and atoms.
    s = replace(s, r";;[^\n]*" => "")
    toks = Any[]
    cs = collect(s)
    i, n = 1, length(cs)
    while i <= n
        c = cs[i]
        if c == '(' || c == ')'
            push!(toks, c == '(' ? :lparen : :rparen); i += 1
        elseif c == '"'
            j = i + 1; buf = IOBuffer()
            while j <= n && cs[j] != '"'
                if cs[j] == '\\' && j + 1 <= n
                    print(buf, cs[j+1]); j += 2
                else
                    print(buf, cs[j]); j += 1
                end
            end
            push!(toks, (:str, String(take!(buf)))); i = j + 1
        elseif isspace(c)
            i += 1
        else
            j = i
            while j <= n && !isspace(cs[j]) && cs[j] != '(' && cs[j] != ')' && cs[j] != '"'
                j += 1
            end
            push!(toks, (:atom, String(cs[i:j-1]))); i = j
        end
    end
    return toks
end

function parse_forms(toks)
    # Returns nested Vectors; strings -> (:str, v), atoms -> (:atom, v).
    pos = Ref(1)
    function rd()
        t = toks[pos[]]; pos[] += 1
        if t === :lparen
            lst = Any[]
            while toks[pos[]] !== :rparen
                push!(lst, rd())
            end
            pos[] += 1
            return lst
        elseif t === :rparen
            error("unexpected )")
        end
        return t  # (:str, ..) or (:atom, ..)
    end
    forms = Any[]
    while pos[] <= length(toks)
        push!(forms, rd())
    end
    return forms
end

# --- query helpers (mirror the Python truthiness: "" counts as absent) ------

isblank(x) = x === nothing || x == ""

head(node) =
    (node isa Vector && !isempty(node) && node[1] isa Tuple) ? node[1][2] : nothing

function findform(node, name)
    # First child list whose head atom == name.
    node isa Vector || return nothing
    for ch in node
        if ch isa Vector && !isempty(ch) && ch[1] isa Tuple && ch[1][2] == name
            return ch
        end
    end
    return nothing
end

first_string(::Nothing) = nothing
function first_string(node)
    # First (non-empty) string literal anywhere under node, depth-first.
    if node isa Tuple
        return node[1] === :str ? node[2] : nothing
    elseif node isa Vector
        for ch in node
            r = first_string(ch)
            if !isblank(r)
                return r
            end
        end
    end
    return nothing
end

function all_strings(node)
    out = String[]
    if node isa Tuple
        node[1] === :str && push!(out, node[2])
    elseif node isa Vector
        for ch in node
            append!(out, all_strings(ch))
        end
    end
    return out
end

# Escape backslash then double-quote, in that order (mirrors the original).
q(s) = "\"" * replace(replace(string(s), "\\" => "\\\\"), "\"" => "\\\"") * "\""

function slug(s)
    t = replace(lowercase(string(s)), r"[^a-z0-9]+" => "-")
    t = strip(t, '-')
    t = first(t, 40)
    return isempty(t) ? "blocker" : String(t)
end

# pick the first truthy (non-blank) candidate, falling back to `dflt`
firsttruthy(dflt, cands...) = (for c in cands; isblank(c) || return c; end; dflt)

# --- conversion -------------------------------------------------------------

function convert(path)
    src = read(path, String)
    forms = parse_forms(tokenize(src))
    state = nothing
    for f in forms
        if head(f) == "state"
            state = f; break
        end
    end
    if state === nothing
        print(stderr, "ERR: no (state ...) form in $path\n")
        exit(3)
    end

    # phase  <- (current-position (phase "..")) | (current-phase "..")
    phase = nothing
    cp = findform(state, "current-position")
    if cp !== nothing
        ph = findform(cp, "phase")
        ph !== nothing && (phase = first_string(ph))
    end
    if isblank(phase)
        cph = findform(state, "current-phase")
        cph !== nothing && (phase = first_string(cph))
    end
    isblank(phase) && (phase = "unknown")

    # next_action <- critical-next-actions, first string (immediate first)
    next_action = "TODO — review and set"
    cna = findform(state, "critical-next-actions")
    if cna !== nothing
        imm = findform(cna, "immediate")
        next_action = firsttruthy(next_action, first_string(imm), first_string(cna))
    end

    # last_action <- last session-history entry's first accomplishment string
    last_action = "migrated from v1 STATE.scm"
    sh = findform(state, "session-history")
    if sh !== nothing
        sessions = [c for c in sh if c isa Vector && head(c) == "session"]
        if !isempty(sessions)
            acc = findform(sessions[end], "accomplishments")
            last_action = firsttruthy(last_action, first_string(acc), first_string(sessions[end]))
        end
    end

    # updated <- (metadata (updated|last-updated "..")) else today
    updated = nothing
    md = findform(state, "metadata")
    if md !== nothing
        for key in ("updated", "last-updated")
            u = findform(md, key)
            if u !== nothing
                updated = first_string(u); break
            end
        end
    end
    isblank(updated) && (updated = string(Dates.today()))

    # blockers <- (blockers-and-issues (critical ..)(high ..)(medium ..)(low ..))
    blockers = Tuple{String,String}[]
    bi = findform(state, "blockers-and-issues")
    if bi !== nothing
        for sev in ("critical", "high", "medium", "low")
            grp = findform(bi, sev)
            if grp !== nothing
                for desc in all_strings(grp)
                    push!(blockers, (sev, desc))
                end
            end
        end
    end

    lines = String[]
    push!(lines, "# SPDX-License-Identifier: AGPL-3.0-or-later")
    push!(lines, "# Migrated from $(basename(path)) by state-scm-to-v2.jl on " *
                 "$(Dates.today())")
    push!(lines, "")
    push!(lines, "@state(version=\"2.0\"):")
    push!(lines, "phase: $(q(phase))")
    push!(lines, "next_action: $(q(next_action))")
    push!(lines, "last_action: $(q(last_action))")
    push!(lines, "updated: $(updated)")
    if !isempty(blockers)
        push!(lines, "")
        push!(lines, "@blockers:")
        for (sev, desc) in blockers
            push!(lines, "- id: $(slug(desc))")
            push!(lines, "  description: $(q(desc))")
            push!(lines, "  waiting_on: $(q(sev * "-priority — internal"))")
            push!(lines, "  since: $(updated)")
        end
        push!(lines, "@end")
    end
    push!(lines, "")
    push!(lines, "@end")
    return join(lines, "\n") * "\n"
end

function main()
    write_mode = "--write" in ARGS
    args = [a for a in ARGS if a != "--write"]
    if length(args) != 1
        print(stderr, "usage: state-scm-to-v2.jl <STATE.scm> [--write]\n")
        exit(2)
    end
    scm = args[1]
    out = convert(scm)
    if write_mode
        dest = joinpath(dirname(scm), "STATE.a2ml")
        open(io -> Base.write(io, out), dest, "w")
        print(stderr, "wrote $dest\n")
    else
        Base.write(stdout, out)
    end
end

main()
